doc/Programming Guide For CppAst.Net.md

```diff
@@ -0,0 +1,534 @@
+In such cases, we would consider using a more "automated" mechanism to solve the problem of generating registration code. This is the approach currently chosen by CE. The overall process of automatically generating reflection code in CE is illustrated in the following figure:
+![](https://markdown-1252066114.cos.ap-guangzhou.myqcloud.com/image20220530192511.png)
+Compared to the conventional `C++` compilation process [as shown in the Normal Mode in the figure], we strategically arrange the compilation pipeline, using `libclang`, to create a secondary `Compile` process as shown on the right side:
+- The first compilation occurs within the tool, processing only header files to extract necessary information, such as class definitions.
+- The second compilation is the actual normal compilation process, incorporating the additional files generated by the tool into the entire compilation.
+Thus, by compiling the tool-generated registration code together with the original code, we can obtain a reflection system with complete runtime information.
+The principle of automatically generating the middleware layer for scripting languages is similar to this scenario. By utilizing this `header-only libclang compile` mode, we can effectively organize the implementation logic of offline tools to complete the required code generation. Doesn't it seem like a promising situation?
+
+---
+## 2. `libclang` - Bringing Hope And Confusion
+
+Before the existence of LLVM libraries, we could only rely on string pattern matching methods like regex to accomplish related tasks. The major drawbacks of this approach were, on one hand, its inefficiency, and on the other hand, the way business programs organized code could disrupt the functioning of automation tools, making the troubleshooting and pinpointing of issues less straightforward. With the advent of LLVM libraries, an increasing number of people started to explore extracting source code information at the AST (Abstract Syntax Tree) level. This approach shifted the resolution of related issues back to C++ itself, which is undoubtedly much more stable and controllable than the previously mentioned string-based mechanisms, with issues being easier to locate and troubleshoot. To utilize this method, let’s first get a basic understanding of libclang.
+
+
+### 2.1 `libclang` And The Changes It Brings
+
+libclang is a part of the LLVM toolchain, and the overall working process of LLVM can be simplified as shown in the diagram below:
+![image.png](https://markdown-1252066114.cos.ap-guangzhou.myqcloud.com/image20240228101434.png)
```

Comment:
All images used by this doc should be part of the doc folder and not link to external sources.

Reply:
Ok, I will try to fix the images later.

Reply:
All done~